### ISSUE TITLE
1.5.6 -> 1.6.0-beta.1 server upgrade error: failed to load any existing snapshots

### ISSUE THREAD

**Issue by Cottand** [hashicorp/nomad#17793]
### Nomad version
1.6.0-beta.1

### Operating system and Environment details

NixOS 22.05 patched with packaged nomad 1.6.0-beta.1
- [Nomad definition](https://github.com/Cottand/nixpkgs/blob/23.05-custom/nixos/modules/services/networking/nomad.nix) (includes systemd config)
- [Nomad 1.6.0.beta-1 declaration](https://github.com/Cottand/nixpkgs/blob/23.05-custom/pkgs/applications/networking/cluster/nomad/default.nix#L81)

### Issue

Nomad agent fails to start up when replacing the binary 1.5.6 -> 1.6.0-beta.1

### Reproduction steps

I attempted both scenarios A and B as described below on 2 different cluster members with the same results every time (both existing servers were 1.5.6 prior to attempting the upgrade)

#### A - Simply replacing the binary
0. I had an existing cluster running (3 servers _and_ clients, 1 client only)
1. Replace Nomad binary on one of the server+client
2. Systemd reload, restart

#### B - Replacing binary + deleting nomad data directories

Not sure to what extent this is not recommended but this is not a production cluster anyway and I thought deleting the existing snapshots in the raft folder might help

0. I had an existing cluster running (3 servers _and_ clients, 1 client only)
1. Systemd stop
2. Replace Nomad binary on one of the server+client
4. Delete all nomad-related non-config folders
5. Systemd reload, restart
6. (Server does seem to discover other peers)

#### Expected Result

Although some features do not work, I thought the new server would at least join the cluster succesfully as explained in [the official guide](https://developer.hashicorp.com/nomad/docs/upgrade#2-check-cluster-health))

#### Actual Result

Server does not start (in scenario A) and prints logs as pasted below (in both scenarios A and B).

In scenario B the server first discovers the cluster (shows up in nomad UI), but then prints the same messages anyway, and fails to start any allocations, but does not crash. If restarted, it does crash like in scenario A.

### Nomad Server logs (if appropriate)

I pasted what I thought was relevant but I can email the entire log if that helps.

```
==> Starting Nomad agent...
==> Error starting agent: server setup failed: Failed to start Raft: failed to load any existing snapshots
    2023-07-04T07:10:43.822+0100 [INFO]  nomad: setting up raft bolt store: no_freelist_sync=false
    2023-07-04T07:10:43.824+0100 [INFO]  nomad.raft: starting restore from snapshot: id=72034-526923-1688450543850 last-index=526923 last-term=72034 size-in-bytes=503363
    2023-07-04T07:10:43.829+0100 [INFO]  nomad.raft: snapshot restore progress: id=72034-526923-1688450543850 last-index=526923 last-term=72034 size-in-bytes=503363 read-bytes=47217 percent-complete="9.38%"
    2023-07-04T07:10:43.829+0100 [ERROR] nomad.raft: failed to restore snapshot: id=72034-526923-1688450543850 last-index=526923 last-term=72034 size-in-bytes=503363 error="job insert failed: missing value for index 'pool'"
    2023-07-04T07:10:43.829+0100 [INFO]  nomad.raft: starting restore from snapshot: id=72034-526923-1688450533215 last-index=526923 last-term=72034 size-in-bytes=503363
    2023-07-04T07:10:43.832+0100 [INFO]  nomad.raft: snapshot restore progress: id=72034-526923-1688450533215 last-index=526923 last-term=72034 size-in-bytes=503363 "read-bytes=47217 percent-complete="9.38%"
    2023-07-04T07:10:43.832+0100 [ERROR] nomad.raft: failed to restore snapshot: id=72034-526923-1688450533215 last-index=526923 last-term=72034 size-in-bytes=503363  error="job insert failed: missing value for index 'pool'
    2023-07-04T07:10:43.832+0100 [ERROR] nomad: failed to start Raft: error="failed to load any existing snapshots"
    2023-07-04T07:10:43.832+0100 [ERROR] agent: error starting agent: error="server setup failed: Failed to start Raft: failed to load any existing snapshots"
```

### Minimal attempted server config:

```hcl
server {
    enabled = true
    bootstrap_expect = 3
    server_join {
        retry_join = [ "10.8.0.1", "10.8.0.5" ]
        retry_max = 3
        retry_interval = "15s"
    }
}

data_dir  = "/var/lib/nomad"

bind_addr = "10.8.0.8"

advertise {
    http = "10.8.0.8"
    rpc  = "10.8.0.8"
    serf = "10.8.0.8" 
}

log_rotate_bytes = 1024000

ports {
    http = 4646
    rpc  = 4647
    serf = 4648
}

telemetry {
    collection_interval = "5s"
    disable_hostname = true
    prometheus_metrics = true
    publish_allocation_metrics = true
    publish_node_metrics = true
}
client {
    enabled = true
    servers = ["10.8.0.1", "10.8.0.5", "10.8.0.8"]

    options = {
        "driver.allowlist" = "docker,raw_exec"
    }

    bridge_network_hairpin_mode = true # only 1.5.+

    host_network "vpn" {
        cidr = "10.8.0.0/24"
        reserved_ports = "51820"
    }

   # a few host volumes (omitted)

    meta {
        box = "ari"
        name = "ari"
        seaweedfs_volume = true
        docker_privileged = true
    }

}
plugin "raw_exec" {
    config {
        enabled = true
    }
}
plugin "docker" {
    config {
        allow_privileged = true
        extra_labels = ["job_name", "task_group_name", "task_name", "node_name"]
    }
}
```

**Comment by Cottand** [comment]
Given the server fails to restore local snapshots too I thought updating all servers at once would make no difference but please let me know if my approach to updating the cluster is wrong and I should not have attempted this in-place one server at a time.

**Comment by jrasell** [comment]
Hi @Cottand, thanks for the detailed report and apologies you hit this bug. It was also reported within https://github.com/hashicorp/nomad/issues/17764 and subsequently fixed within https://github.com/hashicorp/nomad/pull/17765 which will be released as part of the 1.6.0 release candidate. This is within main, if you wanted to try it out before the official release, which should be within the next few weeks.
